casper-deploy-notifier/src/lib.rs

```diff
@@ -0,0 +1,97 @@
+use casper_types::AsymmetricType;
+use eventsource_stream::{Event, EventStreamError, Eventsource};
+use futures::stream::{BoxStream, TryStreamExt};
+use tokio::sync::mpsc;
+
+use crate::error::SseError;
+use crate::sse_types::SseData;
+use crate::types::Notification;
+
+mod error;
+mod sse_types;
+mod types;
+
+const DEFAULT_SSE_SERVER: &str = "https://events.mainnet.casperlabs.io";
+const DEFAULT_EVENT_CHANNEL: &str = "/events/main";
+
+type BoxedEventStream = BoxStream<'static, Result<Event, EventStreamError<reqwest::Error>>>;
+
+pub struct DeployNotifier {
+    url: String,
+}
+
+impl Default for DeployNotifier {
+    fn default() -> Self {
+        let url = format!("{}{}", DEFAULT_SSE_SERVER, DEFAULT_EVENT_CHANNEL);
+        Self { url }
+    }
+}
+
+impl DeployNotifier {
+    pub fn new(url: &str) -> Self {
+        DeployNotifier {
+            url: url.to_string(),
+        }
+    }
+
+    async fn connect(&self) -> Result<BoxedEventStream, SseError> {
+        // Connect to SSE endpoint.
+        let client = reqwest::Client::new();
+        let response = client.get(&self.url).send().await?;
+
+        let stream = response.bytes_stream();
+        let mut event_stream = stream.eventsource();
+
+        // Handle the handshake with API version.
+        let handshake_event = event_stream
+            .try_next()
+            .await?
+            .ok_or(SseError::StreamExhausted)?;
+        let handshake_data: SseData = serde_json::from_str(&handshake_event.data)?;
+        let _api_version = match handshake_data {
+            SseData::ApiVersion(v) => Ok(v),
+            _ => Err(SseError::InvalidHandshake),
+        }?;
+
+        // Wrap stream with box.
+        let boxed_event_stream = Box::pin(event_stream);
+
+        Ok(boxed_event_stream)
+    }
+
+    // Handle incoming events - look for successfuly processed deployments.
+    pub async fn run(&mut self, tx: mpsc::Sender<Notification>) -> Result<(), SseError> {
+        let mut event_stream = self.connect().await?;
+
+        while let Some(event) = event_stream.try_next().await? {
+            let data: SseData = serde_json::from_str(&event.data)?;
+            match data {
+                SseData::ApiVersion(_) => Err(SseError::UnexpectedHandshake)?,
```

Comment:
I don't fully understand why this maps to an `UnexpectedHandshake` error.

Reply:
`ApiVersion` is the first event that sent by server, that is used as the handshake. Once you received it you should not receive it again, thus `UnexpectedHandshake`.

I added clarification in https://github.com/cspr-rad/kairos/pull/40/commits/668faadaecc2a2492409beef8bbd0a2e19b96e4d.